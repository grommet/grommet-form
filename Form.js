'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _FormState = require('./FormState');

var _FormState2 = _interopRequireDefault(_FormState);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Form = function (_Component) {
  _inherits(Form, _Component);

  function Form(props, context) {
    _classCallCheck(this, Form);

    var _this = _possibleConstructorReturn(this, (Form.__proto__ || Object.getPrototypeOf(Form)).call(this, props, context));

    _initialiseProps.call(_this);

    var formState = new _FormState2.default(props.rules, props.defaultValue, _this.onChange);

    _this.state = {
      formState: formState,
      resource: formState.get(),
      errors: props.defaultErrors,
      rules: props.rules,
      isFormSubmitted: false
    };
    return _this;
  }

  _createClass(Form, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.reset || !(0, _utils.deepCompare)(nextProps.rules, this.props.rules) || !(0, _utils.deepCompare)(nextProps.defaultErrors, this.props.defaultErrors) || !(0, _utils.deepCompare)(nextProps.defaultValue, this.props.defaultValue)) {
        this.reset(nextProps);
      }
    }
  }, {
    key: 'reset',
    value: function reset(props) {
      var formState = new _FormState2.default(props.rules, props.defaultValue, this.onChange);

      this.setState({
        formState: formState,
        resource: formState.get(),
        errors: props.defaultErrors,
        rules: props.rules
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var children = this.props.children;
      var _state = this.state,
          _state$errors = _state.errors,
          errors = _state$errors === undefined ? {} : _state$errors,
          formState = _state.formState,
          resource = _state.resource,
          rules = _state.rules;

      var state = this.getObjectStateValue(typeof rules === 'function' ? rules(resource) : rules);
      state.get = function (key) {
        return _this2.getStateValue(key);
      };
      state.set = function (key, value) {
        return _this2.setStateValue(key, value);
      };
      errors.get = function (key) {
        return (0, _utils.getValueByKey)(errors, key);
      };
      return _react2.default.createElement(
        'form',
        { onSubmit: this.onSubmit },
        children(state, errors, formState.isValid())
      );
    }
  }]);

  return Form;
}(_react.Component);

Form.defaultProps = {
  defaultErrors: undefined,
  defaultValue: undefined,
  onError: undefined,
  onChange: undefined,
  reset: undefined
};
Form.propTypes = {
  children: _propTypes2.default.func.isRequired,
  defaultErrors: _propTypes2.default.object,
  defaultValue: _propTypes2.default.object,
  onSubmit: _propTypes2.default.func.isRequired,
  onError: _propTypes2.default.func,
  onChange: _propTypes2.default.func,
  reset: _propTypes2.default.bool,
  rules: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.func]).isRequired
};

var _initialiseProps = function _initialiseProps() {
  var _this3 = this;

  this.onChange = function (updated) {
    var _state2 = _this3.state,
        isFormSubmitted = _state2.isFormSubmitted,
        formState = _state2.formState;

    if (isFormSubmitted) {
      _this3.setState({ resource: formState.get(), errors: formState.getErrors() });
    } else {
      _this3.setState({ resource: formState.get(), errors: undefined });
    }
    if (_this3.props.onChange) {
      _this3.props.onChange(updated);
    }
  };

  this.onSubmit = function (event) {
    var _props = _this3.props,
        onSubmit = _props.onSubmit,
        onError = _props.onError;
    var formState = _this3.state.formState;

    event.preventDefault();
    if (formState.isValid()) {
      onSubmit(formState.getResource(), function () {
        return _this3.reset(_this3.props);
      });
    } else {
      var errors = formState.getErrors();
      if (typeof onError === 'function') {
        onError(errors);
      }
      _this3.setState({ isFormSubmitted: true, errors: errors });
    }
  };

  this.getStateValue = function (key) {
    var _state3 = _this3.state,
        formState = _state3.formState,
        resource = _state3.resource;

    return {
      name: key,
      value: (0, _utils.getValueByKey)(resource, key) || '',
      checked: (0, _utils.getValueByKey)(resource, key) === true,
      onChange: function onChange(event) {
        var value = typeof event === 'string' ? event : event.option;
        if (!value && event.target) {
          value = event.target.value;
          if (event.target && event.target.type === 'checkbox' && (event.target.value === '' || event.target.value === 'true')) {
            value = event.target.checked;
          }
        }
        formState.set(key, typeof value === 'undefined' ? '' : value);
      }
    };
  };

  this.setStateValue = function (key, value) {
    var formState = _this3.state.formState;

    formState.set(key, value);
  };

  this.getObjectStateValue = function (obj, parentKey) {
    var objStateValue = {};
    Object.keys(obj).forEach(function (key) {
      if (typeof obj[key] === 'string' || typeof obj[key] === 'function') {
        objStateValue[key] = _this3.getStateValue(parentKey ? parentKey + '.' + key : key);
      } else if ((0, _utils.isObject)(obj[key])) {
        objStateValue[key] = _this3.getObjectStateValue(obj[key], parentKey ? parentKey + '.' + key : key);
      }
    });
    return objStateValue;
  };
};

exports.default = Form;