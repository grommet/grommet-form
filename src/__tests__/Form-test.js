import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Form from '../Form';

Enzyme.configure({ adapter: new Adapter() });

const emailExpression = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

const rules = {
  email: (value) => {
    if (!value || value === '') {
      return 'Email is required';
    } else if (!emailExpression.test(value)) {
      return 'Email is not valid';
    }
    return undefined;
  },
  name: 'Name is required',
  address: {
    street: 'Street is required',
  },
};

test('Form renders', () => {
  const onSubmit = jest.fn();
  const defaultProps = {
    defaultErrors: { email: 'This is an existing email' },
    defaultValue: { email: 'alan@gmail.com' },
    onSubmit,
    rules,
  };
  const component = mount(
    <Form {...defaultProps}>
      {(state, errors) => (
        <fieldset>
          <input {...state.name} />
          {errors.name ? <span className='error'>{errors.get('name')}</span> : undefined}
          <input {...state.email} />
          {errors.email ? <span className='error'>{errors.email}</span> : undefined}
          <input {...state.get('address.street')} />
          {errors.get('address.street') ? (
            <span className='error'>{errors.address.street}</span>
          ) : undefined}
          <input {...state.get('optional.field')} />
          <button type='submit'>Add</button>
          <button
            type='button'
            onClick={() => {
              state.set('name', 'test_default');
              state.set({
                email: 'test@default.com',
              });
            }}
          >
            Set Default
          </button>
        </fieldset>
      )}
    </Form>
  );

  expect(component.getDOMNode()).toMatchSnapshot();

  component.find('input').forEach(node => node.simulate('change', { target: { value: '' } }));

  component.find("button[type='submit']").simulate('submit');

  expect(component.getDOMNode()).toMatchSnapshot();

  component.setProps({ defaultErrors: { name: 'This is a name error' } });

  expect(component.getDOMNode()).toMatchSnapshot();

  component.setProps(
    { rules: () => ({ name: 'just name is required' }), defaultErrors: undefined }
  );
  component.find("button[type='submit']").simulate('submit');

  expect(component.getDOMNode()).toMatchSnapshot();

  component.setProps({ defaultValue: { name: 'default name' } });

  expect(component.getDOMNode()).toMatchSnapshot();

  component.setProps(defaultProps);

  expect(component.getDOMNode()).toMatchSnapshot();

  component.find('input').forEach((node, index) => {
    if (index === 1) {
      node.simulate('change', { target: { value: 'test@gmail.com' } });
    } else {
      node.simulate('change', { target: { value: 'test' } });
    }
  });

  component.find("button[type='submit']").simulate('submit');
  expect(component.getDOMNode()).toMatchSnapshot();
  expect(onSubmit).toBeCalledWith(
    {
      address: { street: 'test' },
      email: 'test@gmail.com',
      name: 'test',
      optional: { field: 'test' } },
  );

  component.find("button[type='button']").simulate('click');
  component.find("button[type='submit']").simulate('submit');
  expect(component.getDOMNode()).toMatchSnapshot();
  expect(onSubmit).toBeCalledWith(
    {
      address: { street: 'test' },
      email: 'test@default.com',
      name: 'test_default',
      optional: { field: 'test' } },
  );
});
