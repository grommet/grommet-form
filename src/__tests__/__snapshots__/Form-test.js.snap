// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders 1`] = `
<form>
  <fieldset>
    <input
      name="name"
      value=""
    />
    <input
      name="email"
      value="alan@gmail.com"
    />
    <span
      class="error"
    >
      This is an existing email
    </span>
    <input
      name="address.street"
      value=""
    />
    <input
      name="optional.field"
      value=""
    />
    <button
      type="submit"
    >
      Add
    </button>
    <button
      type="button"
    >
      Set Default
    </button>
  </fieldset>
</form>
`;

exports[`Form renders 2`] = `
<form>
  <fieldset>
    <input
      name="name"
      value=""
    />
    <span
      class="error"
    >
      Name is required
    </span>
    <input
      name="email"
      value=""
    />
    <span
      class="error"
    >
      Email is required
    </span>
    <input
      name="address.street"
      value=""
    />
    <span
      class="error"
    >
      Street is required
    </span>
    <input
      name="optional.field"
      value=""
    />
    <button
      type="submit"
    >
      Add
    </button>
    <button
      type="button"
    >
      Set Default
    </button>
  </fieldset>
</form>
`;

exports[`Form renders 3`] = `
<form>
  <fieldset>
    <input
      name="name"
      value=""
    />
    <span
      class="error"
    >
      This is a name error
    </span>
    <input
      name="email"
      value=""
    />
    <input
      name="address.street"
      value=""
    />
    <input
      name="optional.field"
      value=""
    />
    <button
      type="submit"
    >
      Add
    </button>
    <button
      type="button"
    >
      Set Default
    </button>
  </fieldset>
</form>
`;

exports[`Form renders 4`] = `
<form>
  <fieldset>
    <input
      name="name"
      value=""
    />
    <span
      class="error"
    >
      just name is required
    </span>
    <input
      value="alan@gmail.com"
    />
    <input
      name="address.street"
      value=""
    />
    <input
      name="optional.field"
      value=""
    />
    <button
      type="submit"
    >
      Add
    </button>
    <button
      type="button"
    >
      Set Default
    </button>
  </fieldset>
</form>
`;

exports[`Form renders 5`] = `
<form>
  <fieldset>
    <input
      name="name"
      value="default name"
    />
    <input
      value="alan@gmail.com"
    />
    <input
      name="address.street"
      value=""
    />
    <input
      name="optional.field"
      value=""
    />
    <button
      type="submit"
    >
      Add
    </button>
    <button
      type="button"
    >
      Set Default
    </button>
  </fieldset>
</form>
`;

exports[`Form renders 6`] = `
<form>
  <fieldset>
    <input
      name="name"
      value=""
    />
    <input
      value=""
      name="email"
    />
    <span
      class="error"
    >
      This is an existing email
    </span>
    <input
      name="address.street"
      value=""
    />
    <input
      name="optional.field"
      value=""
    />
    <button
      type="submit"
    >
      Add
    </button>
    <button
      type="button"
    >
      Set Default
    </button>
  </fieldset>
</form>
`;

exports[`Form renders 7`] = `
<form>
  <fieldset>
    <input
      name="name"
      value="test"
    />
    <input
      value="test@gmail.com"
      name="email"
    />
    <input
      name="address.street"
      value="test"
    />
    <input
      name="optional.field"
      value="test"
    />
    <button
      type="submit"
    >
      Add
    </button>
    <button
      type="button"
    >
      Set Default
    </button>
  </fieldset>
</form>
`;

exports[`Form renders 8`] = `
<form>
  <fieldset>
    <input
      name="name"
      value="test_default"
    />
    <input
      value="test@default.com"
      name="email"
    />
    <input
      name="address.street"
      value="test"
    />
    <input
      name="optional.field"
      value="test"
    />
    <button
      type="submit"
    >
      Add
    </button>
    <button
      type="button"
    >
      Set Default
    </button>
  </fieldset>
</form>
`;

exports[`Form renders with empty string and checkbox 1`] = `
<form>
  <fieldset>
    <input
      name="email"
      value=""
    />
    <input
      name="name"
      value=""
    />
    <input
      type="checkbox"
      name="nice"
      value=""
    />
    <input
      type="radiobutton"
      name="gender"
      value=""
    />
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;

exports[`Form works with Grommet DateTime 1`] = `
<form>
  <fieldset>
    <div
      class="grommetux-date-time"
      id="-1887354600"
    >
      <input
        name="date"
        class="grommetux-input grommetux-date-time__input"
        placeholder="M/D/YYYY h:mm a"
        value=""
      />
      <button
        type="button"
        class="grommetux-button grommetux-button--plain grommetux-date-time__control"
        aria-label="Date Time Icon"
      >
        <span
          class="grommetux-button__icon"
        >
          <svg
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            height="24px"
            role="img"
            class="grommetux-control-icon grommetux-control-icon-clock grommetux-control-icon--responsive"
            aria-label="clock"
          >
            <path
              fill="none"
              stroke="#000"
              stroke-width="2"
              d="M12,22 C17.5228475,22 22,17.5228475 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,17.5228475 6.4771525,22 12,22 Z M12,5 L12,12 L17,12"
            />
          </svg>
        </span>
      </button>
    </div>
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;

exports[`Form works with Grommet DateTime 2`] = `
<form>
  <fieldset>
    <div
      class="grommetux-date-time"
      id="-1887354600"
    >
      <input
        name="date"
        class="grommetux-input grommetux-date-time__input"
        placeholder="M/D/YYYY h:mm a"
        value=""
      />
      <button
        type="button"
        class="grommetux-button grommetux-button--plain grommetux-date-time__control"
        aria-label="Date Time Icon"
      >
        <span
          class="grommetux-button__icon"
        >
          <svg
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            height="24px"
            role="img"
            class="grommetux-control-icon grommetux-control-icon-clock grommetux-control-icon--responsive"
            aria-label="clock"
          >
            <path
              fill="none"
              stroke="#000"
              stroke-width="2"
              d="M12,22 C17.5228475,22 22,17.5228475 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,17.5228475 6.4771525,22 12,22 Z M12,5 L12,12 L17,12"
            />
          </svg>
        </span>
      </button>
    </div>
    <span
      class="error"
    >
      is required
    </span>
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;

exports[`Form works with Grommet DateTime 3`] = `
<form>
  <fieldset>
    <div
      class="grommetux-date-time"
      id="-1887354600"
    >
      <input
        name="date"
        class="grommetux-input grommetux-date-time__input"
        placeholder="M/D/YYYY h:mm a"
        value="11/20/2017"
      />
      <button
        type="button"
        class="grommetux-button grommetux-button--plain grommetux-date-time__control"
        aria-label="Date Time Icon"
      >
        <span
          class="grommetux-button__icon"
        >
          <svg
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            height="24px"
            role="img"
            class="grommetux-control-icon grommetux-control-icon-clock grommetux-control-icon--responsive"
            aria-label="clock"
          >
            <path
              fill="none"
              stroke="#000"
              stroke-width="2"
              d="M12,22 C17.5228475,22 22,17.5228475 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,17.5228475 6.4771525,22 12,22 Z M12,5 L12,12 L17,12"
            />
          </svg>
        </span>
      </button>
    </div>
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;

exports[`Form works with Grommet DateTime 4`] = `
<form>
  <fieldset>
    <div
      class="grommetux-date-time"
      id="-1887354600"
    >
      <input
        name="date"
        class="grommetux-input grommetux-date-time__input"
        placeholder="M/D/YYYY h:mm a"
        value="11/20/2017"
      />
      <button
        type="button"
        class="grommetux-button grommetux-button--plain grommetux-date-time__control"
        aria-label="Date Time Icon"
      >
        <span
          class="grommetux-button__icon"
        >
          <svg
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            height="24px"
            role="img"
            class="grommetux-control-icon grommetux-control-icon-clock grommetux-control-icon--responsive"
            aria-label="clock"
          >
            <path
              fill="none"
              stroke="#000"
              stroke-width="2"
              d="M12,22 C17.5228475,22 22,17.5228475 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,17.5228475 6.4771525,22 12,22 Z M12,5 L12,12 L17,12"
            />
          </svg>
        </span>
      </button>
    </div>
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;

exports[`Form works with Grommet Select 1`] = `
<form>
  <fieldset>
    <div
      class="grommetux-select"
    >
      <input
        type="text"
        name="size"
        class="grommetux-input grommetux-select__input"
        readonly=""
        value=""
      />
      <button
        type="button"
        class="grommetux-button grommetux-button--plain grommetux-select__control"
        aria-label="Select Icon"
      >
        <span
          class="grommetux-button__icon"
        >
          <svg
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            height="24px"
            role="img"
            class="grommetux-control-icon grommetux-control-icon-caret-down grommetux-control-icon--responsive"
            aria-label="caret-down"
          >
            <polygon
              fill="none"
              stroke="#000"
              stroke-width="2"
              points="22 8 12 20 2 8"
            />
          </svg>
        </span>
      </button>
    </div>
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;

exports[`Form works with Grommet Select 2`] = `
<form>
  <fieldset>
    <div
      class="grommetux-select"
    >
      <input
        type="text"
        name="size"
        class="grommetux-input grommetux-select__input"
        readonly=""
        value=""
      />
      <button
        type="button"
        class="grommetux-button grommetux-button--plain grommetux-select__control"
        aria-label="Select Icon"
      >
        <span
          class="grommetux-button__icon"
        >
          <svg
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            height="24px"
            role="img"
            class="grommetux-control-icon grommetux-control-icon-caret-down grommetux-control-icon--responsive"
            aria-label="caret-down"
          >
            <polygon
              fill="none"
              stroke="#000"
              stroke-width="2"
              points="22 8 12 20 2 8"
            />
          </svg>
        </span>
      </button>
    </div>
    <span
      class="error"
    >
      is required
    </span>
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;

exports[`Form works with Grommet Select 3`] = `
<div
  class="grommet grommetux-drop"
  style="left: 0px; width: 0.1px; top: 0px; max-height: 768px;"
>
  <div>
    <div
      class="grommetux-select__drop"
    >
      <ol
        class="grommetux-select__options"
      >
        <li
          class="grommetux-select__option"
        >
          one
        </li>
        <li
          class="grommetux-select__option"
        >
          two
        </li>
        <li
          class="grommetux-select__option"
        >
          three
        </li>
      </ol>
    </div>
  </div>
</div>
`;

exports[`Form works with Grommet Select 4`] = `
<form>
  <fieldset>
    <div
      class="grommetux-select"
      id="1894311200"
    >
      <input
        type="text"
        name="size"
        class="grommetux-input grommetux-select__input"
        readonly=""
        value="two"
      />
      <button
        type="button"
        class="grommetux-button grommetux-button--plain grommetux-select__control"
        aria-label="Select Icon"
      >
        <span
          class="grommetux-button__icon"
        >
          <svg
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            height="24px"
            role="img"
            class="grommetux-control-icon grommetux-control-icon-caret-down grommetux-control-icon--responsive"
            aria-label="caret-down"
          >
            <polygon
              fill="none"
              stroke="#000"
              stroke-width="2"
              points="22 8 12 20 2 8"
            />
          </svg>
        </span>
      </button>
    </div>
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;

exports[`Form works with Grommet Select 5`] = `
<form>
  <fieldset>
    <div
      class="grommetux-select"
      id="1894311200"
    >
      <input
        type="text"
        name="size"
        class="grommetux-input grommetux-select__input"
        readonly=""
        value="two"
      />
      <button
        type="button"
        class="grommetux-button grommetux-button--plain grommetux-select__control"
        aria-label="Select Icon"
      >
        <span
          class="grommetux-button__icon"
        >
          <svg
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            height="24px"
            role="img"
            class="grommetux-control-icon grommetux-control-icon-caret-down grommetux-control-icon--responsive"
            aria-label="caret-down"
          >
            <polygon
              fill="none"
              stroke="#000"
              stroke-width="2"
              points="22 8 12 20 2 8"
            />
          </svg>
        </span>
      </button>
    </div>
    <button
      type="submit"
    >
      Add
    </button>
  </fieldset>
</form>
`;
